Certificate Distribution Using Ansbile From Linux SP Server to Linux

msankpal92 opened the pull request:
…SP Clients

## PR summary

Added the Ansible playbook to distribute the IBM Storage Protect Server certificate to clients.

**Fixes:** <! -- link to issue -->

## PR Checklist
Please make sure that your PR fulfills the following requirements:  
- [ ] The commit message follows the Commit Message Guidelines in the CONTRIBUTING.md.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type  
<!-- Please check the one that applies to this PR using "x". -->
- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [ ] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## What is the current behavior?  
NA

## What is the new behavior?  

The proposed Ansbile playbook has multiple plays running on different set of hosts for distributing IBM Storage Protect server certificate to its connected clients.

## Does this PR introduce a breaking change?    
- [ ] Yes
- [ ] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information
Added a new folder named playbook containing certificate distribute folder where along with the playbook other files are also present. File data.ini contains the variables that will be used during distribution of certificate. CFG inventory file contains the IBM Storage Protect server host information along with ADMIN client and rest of the clients host information.

Tompage1994 commented on playbooks/cert-distribute/cert_distribute.yml:
```diff
@@ -0,0 +1,69 @@
+        var: sp_server_date_time.stdout
+
+    - name: Create a new self-signed certificate on the IBM Storage Protect server.
+      shell: "cd /opt/tivoli/tsm/client/ba/bin/;dsmadmc -id={{ lookup('ini', 'SP_SERVER_USERNAME', file='data.ini') }} -password={{ lookup('ini', 'SP_SERVER_PASSWORD', file='data.ini') }} \"create certificate \\\"{{ sp_server_date_time.stdout }}\\\"\""
```
It feels like this could be a module in itself.

Either way, could this be changed as follows:

```suggestion
      command: 
        chdir: /opt/tivoli/tsm/client/ba/bin/
        cmd: "dsmadmc -id={{ lookup('ini', 'SP_SERVER_USERNAME', file='data.ini') }} -password={{ lookup('ini', 'SP_SERVER_PASSWORD', file='data.ini') }} \"create certificate \\\"{{ sp_server_date_time.stdout }}\\\"\""
```

Tompage1994 reviewed (changes requested):
One key change which needs to be made is that the name of the directory should be `cert_distribute` otherwise there will be problems trying to import the playbook

Two comments:
1. There is a lot of `shell` module in here. As much as possible we should look to reduce this.
2. Playbooks in the `playbooks` directory of a collection should be fully runable without editing anything (and only providing an inventory). This needs to be refactored to allow this type of execution such that one could run:
```
- ansible.builtin.import_playbook:
    name: ibm.storage_protect.cert_distribute
```

msankpal92 commented:
Hi @Tompage1994 , 

Renamed the folder and updated the changes you have suggested.

About this certificate distribution task :

1. We generate certificate on IBM Storage Protect Server. It gets generated in servers instance directory.
2. We execute above task using ADMIN client.
3. We copy from there to local where we are running our playbook
4. Finally we distribute it to all clients on a specific directory.

So here, total hosts are : 

1. IBM Storage Protect Server
2. ADMIN Client (This can be local)
3. Clients connected to server

Also, here this playbook can be used multiple times to distribute certificate from a specific server to its connected clients.

So, instead of reading values from environment variable, I did put them into data.ini for better use.

Thank You.

msankpal92 commented:
> One key change which needs to be made is that the name of the directory should be `cert_distribute` otherwise there will be problems trying to import the playbook
> 
> Two comments:
> 
> 1. There is a lot of `shell` module in here. As much as possible we should look to reduce this.
> 2. Playbooks in the `playbooks` directory of a collection should be fully runable without editing anything (and only providing an inventory). This needs to be refactored to allow this type of execution such that one could run:
> 
> ```
> - ansible.builtin.import_playbook:
>     name: ibm.storage_protect.cert_distribute
> ```

Changes from 3 commits rebased into 1 commit for resolving DCO validation

Tompage1994 commented:
> 
> So, instead of reading values from environment variable, I did put them into data.ini for better use.
> 

Ok, that's fine, but the file that's there isn't useful to anyone. It would need to be generic as people will NOT be able to change these ini files because they will include this playbook from their own.

msankpal92 commented:
> Ok, that's fine, but the file that's there isn't useful to anyone. It would need to be generic as people will NOT be able to change these ini files because they will include this playbook from their own.

Hi @Tompage1994 ,

This certificate distribution from IBM Storage Protect server to all its connected clients will be executed multiple times. For different pair of server clients.

These files are for setting up details of server and its administrator username and password. Which is different from the machine credentials. So, these files will be useful to end users to directly edit and use them in same format. Without much to edit in the playbook code.

Apart from this, we have developed this automation for Windows - PowerShell, Linux - Shell, and AIX - KSH using scripts. Where apart from script other configuration files are there to edit. So, the idea is to keep similar behaviour when working with Ansible.
https://www.ibm.com/docs/en/storage-protect/8.1.25?topic=errors-automating-distribution-storage-protect-server-certificate-clients

Thank You.
Mahesh

msankpal92 commented:
**Unit test demo clip :**


https://github.com/user-attachments/assets/5be4d336-83e6-4ab4-b9cb-a1e71139756f
